### PR TITLE
Fix AD/DC auth inheriting issue which caused sign-out not clean up

### PR DESCRIPTION
**Pull request by wezhang**

### CODE REVIEW

File: Utils/azuretools-core/src/com/microsoft/azuretools/authmanage/AuthMethodManager.java [hunk 1]
```diff
@@ -171,13 +179,25 @@ private synchronized AzureManager getAzureManager(final AuthMethod authMethod) t
         return localAzureManagerRef;
     }
 
+    public BaseADAuthManager getAdAuthManagerBy(final AuthMethod authMethod) {
```
Comment (Flanker32) [inline Utils/azuretools-core/src/com/microsoft/azuretools/authmanage/AuthMethodManager.java]: change to `getAdAuthManagerByAuthMethod`?

Comment (wezhang) [reply Utils/azuretools-core/src/com/microsoft/azuretools/authmanage/AuthMethodManager.java]: Fix it!

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.core/src/com/microsoft/azuretools/core/ui/SignInDialog.java [hunk 2]
```diff
@@ -321,27 +333,26 @@ private void signInAsync(final DCAuthManager dcAuthManager) throws InvocationTar
 
     private void doSignOut() {
         accountEmail = null;
-        DCAuthManager.getInstance().signOut();
+        // AuthMethod.AD is deprecated.
+        getAuthMethodManager().getAdAuthManagerByAuthMethod(AuthMethod.DC).signOut();
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.core/src/com/microsoft/azuretools/core/ui/SignInDialog.java]: Shall we use `getAuthMethodManager().signOut();` here?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azuretools/ijidea/ui/SignInWindow.java [hunk 3]
```diff
@@ -220,30 +229,29 @@ public void run(ProgressIndicator indicator) {
     private void doSignOut() {
         try {
             accountEmail = null;
-            AdAuthManager.getInstance().signOut();
+            // AuthMethod.AD is deprecated.
+            getAuthMethodManager().getAdAuthManagerByAuthMethod(AuthMethod.DC).signOut();
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azuretools/ijidea/ui/SignInWindow.java]: Same as above, could we call singout of getAuthMethodManager() directly?

File: Utils/azuretools-core/src/com/microsoft/azuretools/authmanage/AuthMethodManager.java [hunk 4]
```diff
@@ -132,23 +137,26 @@ private synchronized AzureManager getAzureManager(final AuthMethod authMethod) t
         AzureManager localAzureManagerRef = azureManager;
 
         if (localAzureManagerRef == null) {
-            final String accountEmail = authMethodDetails.getAccountEmail();
             switch (authMethod) {
                 case AD:
-                    if (StringUtils.isNullOrEmpty(accountEmail) ||
-                            (!AdAuthManager.getInstance().isSignedIn() &&
-                                    !AdAuthManager.getInstance().tryRestoreSignIn(authMethodDetails))) {
+                    if (isBlank(authMethodDetails.getAccountEmail())) {
```
Comment (Flanker32) [inline Utils/azuretools-core/src/com/microsoft/azuretools/authmanage/AuthMethodManager.java]: Shall we merge the AD and DC parts, seems some of the codes are duplicated

File: Utils/azuretools-core/src/com/microsoft/azuretools/authmanage/AuthMethodManager.java [hunk 5]
```diff
@@ -200,35 +224,56 @@ public AuthMethodDetails getAuthMethodDetails() {
...
+            return JsonHelper.deserialize(AuthMethodDetails.class, json);
         } catch (IOException ignored) {
             System.out.println("Failed to loading authMethodDetails settings. Use defaults.");
-            authMethodDetails = new AuthMethodDetails();
+            return new AuthMethodDetails();
+        }
+    }
+
+    private void restoreADSignIn() {
+        switch (getAuthMethod()) {
```
Comment (Flanker32) [inline Utils/azuretools-core/src/com/microsoft/azuretools/authmanage/AuthMethodManager.java]: It seems that we could use a common method for AD/DC

File: Utils/azuretools-core/src/com/microsoft/azuretools/authmanage/BaseADAuthManager.java [hunk 6]
```diff
@@ -23,24 +23,31 @@
...
 import com.microsoft.azuretools.service.ServiceManager;
+
 import java.io.IOException;
 import java.util.UUID;
 import java.util.logging.Logger;
 
+import static org.apache.commons.lang3.StringUtils.isNoneBlank;
+
 public abstract class BaseADAuthManager {
+    protected static final String AUTHORIZATIONREQUIRED = "Authorization is required, please sign out and sign in again";
```
Comment (Flanker32) [inline Utils/azuretools-core/src/com/microsoft/azuretools/authmanage/BaseADAuthManager.java]: change to `AUTHORIZATION_REQUIRED`?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azuretools/ijidea/ui/SignInWindow.java (hunk 3 above)
Comment (wezhang) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azuretools/ijidea/ui/SignInWindow.java]: Looks like the current method is only for DC auth.

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.core/src/com/microsoft/azuretools/core/ui/SignInDialog.java (hunk 2 above)
Comment (wezhang) [reply PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.core/src/com/microsoft/azuretools/core/ui/SignInDialog.java]: Looks like the current method is only for DC auth.

File: Utils/azuretools-core/src/com/microsoft/azuretools/authmanage/AuthMethodManager.java (hunk 4 above)
Comment (wezhang) [reply Utils/azuretools-core/src/com/microsoft/azuretools/authmanage/AuthMethodManager.java]: fix it.

File: Utils/azuretools-core/src/com/microsoft/azuretools/authmanage/AuthMethodManager.java (hunk 5 above)
Comment (wezhang) [reply Utils/azuretools-core/src/com/microsoft/azuretools/authmanage/AuthMethodManager.java]: fix it.

File: Utils/azuretools-core/src/com/microsoft/azuretools/authmanage/BaseADAuthManager.java (hunk 6 above)
Comment (wezhang) [reply Utils/azuretools-core/src/com/microsoft/azuretools/authmanage/BaseADAuthManager.java]: fix it.

**Pull request merged**